Sources/SwiftSyntax/BumpPtrAllocator.swift

```diff
@@ -0,0 +1,113 @@
+//===---------- BumpPtrAllocator.swift - Bump-pointer Allocaiton ----------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+/// Bump-pointer allocation.
+class BumpPtrAllocator {
+  typealias Slab = UnsafeMutableRawBufferPointer
+
+  static private var SLAB_SIZE: Int = 4096
+  static private var GLOWTH_DELAY: Int = 128
+  static private var SLAB_ALIGNMENT: Int = 8
+
+  private var slabs: [Slab]
+  private var currentPtr: UnsafeMutableRawPointer?
+  private var endPtr: UnsafeMutableRawPointer?
```

Comment:
Could you add comments to these?

```suggestion
  /// Points to the next unused address in `slabs.last`
  private var currentPtr: UnsafeMutableRawPointer?
  /// Points to the end address of `slabs.last`. If `slabs` is not empty, this is equivalent to
  /// `slabs.last!.baseAddress! + slabs.last!.count`
  private var endPtr: UnsafeMutableRawPointer?
```

Also, I wonder if it makes sense to start an initial slab when the bump allocator is created so these can be non-optional. I think the use case of creating a bump allocator but never allocating anything in it are very rare.

Reply:
:+1: for the comments.
Two reasons I didn't make these non-optional:
* Need to allocate the first slab _before_ initializing these, that means I cannot use `startNewSlab` for the first slab
* As I commented above, I want to reset these before deallocating the first slab

Reply:
Fair enough 👍 I’m also fine to keep it as-is.